File: pytoaq/latin.py
# -*- coding: utf-8 -*-

# LATIN TOAQ MODULE

# ==================================================================== #

import regex as re, unicodedata

# ==================================================================== #

vowel_str = "aeiıouáéíóúäëïöüâêîôûạẹịı̣ọụạ́ẹ́ị́ọ́ụ́ạ̈ẹ̈ị̈ọ̈ụ̈ậệị̂ộụ̂"
std_vowel_str = "aeıouáéíóúäëïöüâêîôûạẹı̣ọụạ́ẹ́ị́ọ́ụ́ạ̈ẹ̈ị̈ọ̈ụ̈ậệị̂ộụ̂"
vowels = vowel_str
std_vowels = std_vowel_str
consonant_str = "'bcdfghjȷklmnprstzqꝡ"
initial_str = "'bcdfghjȷklmnprstzꝡ"
word_initial_str = "bcdfghjȷklmnprstzꝡ"
std_consonant_str = "'bcdfghjklmnprstzqꝡ"
std_initial_str = "'bcdfghjklmnprstzꝡ"
std_word_initial_str = "bcdfghjklmnprstzꝡ"
charset = vowel_str + consonant_str
std_charset = std_vowel_str + std_consonant_str

initials = ("m", "b", "p", "f", "n", "d", "t", "z", "c", "s", "r", "l", "nh", "j", "ȷ", "ch", "sh", "ꝡ", "g", "k", "'", "h")
finals = ("m", "q")
consonants = initials + ("q",)
std_initials = ("m", "b", "p", "f", "n", "d", "t", "z", "c", "s", "r", "l", "nh", "j", "ch", "sh", "ꝡ", "g", "k", "'", "h")
std_consonants = std_initials + ("q",)

# ==================================================================== #

root_subordinators = {"ꝡa", "ma", "tıo"}
nominal_subordinators = {"ꝡä", "mä", "tïo", "lä", "ꝡé", "ná", "é"}
adnominal_subordinators = {"ꝡë", "ë", "jü"}
predicatizers = {"jeı", "mea", "po"}
determiners = {"ló", "ké", "sá", "sía", "tú", "túq", "báq", "já", "hí", "ní", "hú"}
type_1_conjunctions = {"róı", "rú", "rá", "ró", "rí", "kéo"}
type_2_conjunctions = {"roı", "ru", "ra", "ro", "rı", "keo"}
type_3_conjunctions = {"rôı", "rû", "râ", "rô", "rî", "kêo"}
conjunctions = (
  type_1_conjunctions | type_2_conjunctions | type_3_conjunctions)
falling_tone_illocutions = {"ka", "da", "ba", "nha", "doa", "ꝡo"}
peaking_tone_illocutions = {"dâ", "môq"}
raising_tone_illocutions = {"móq"}
illocutions = (
  falling_tone_illocutions | raising_tone_illocutions
  | peaking_tone_illocutions)
focus_markers = {"kú", "tóu", "béı"}
clefts = {"bï", "nä", "gö"}
vocative = {"hóı"}
terminators = {"teo", "kı"}
exophoric_pronouns = {
  "jí", "súq", "nháo", "súna", "nhána", "úmo", "íme", "súho", "áma", "há"
}
endophoric_pronouns = {"hóa", "hó", "máq", "tá", "hóq", "róu", "zé", "bóu", "áq", "chéq"}
pronouns = exophoric_pronouns | endophoric_pronouns

functors_with_grammatical_tone = predicatizers | {"mı", "shu", "mo"}

functors_with_lexical_tone = (
  root_subordinators | nominal_subordinators | adnominal_subordinators
  | determiners | conjunctions | illocutions | focus_markers
  | clefts | vocative | terminators | {"kïo"})

functor_lemmas = (
  functors_with_grammatical_tone | functors_with_lexical_tone)

toneless_particles = (
  root_subordinators | falling_tone_illocutions | terminators)

interjections = {
  'm̄', 'ḿ', 'm̈', 'm̉', 'm̂', 'm̀', 'm̃'
}

# ==================================================================== #

def is_a_word(s):
  raise NotImplementedError()

def _with_replaced_characters(str, src_chars, rep_chars):
  i = 0
  while i < len(str):
    if str[i] in src_chars:
      str = str[:i] + rep_chars[src_chars.index(str[i])] + str[i+1:]
    i += 1
  return str

def diacriticless_normalized(s):
  s = s.lower()
  s = re.sub("[x’]", "'", s)
  s = re.sub("(?<=^)'", "", s)
  s = unicodedata.normalize("NFD", s)
  #s = re.sub("[^0-9A-Za-zı\u0300-\u030f'_ ()«»,;.…!?]+", " ", s)
  s = re.sub("[^0-9A-Za-zı'_ ()«»,;.…!?]+", "", s)
  s = unicodedata.normalize("NFC", s)
  s = re.sub("i", "ı", s)
  s = re.sub("ȷ", "j", s)
  return s.strip()
  
def lemma_of(s):
  return diacriticless_normalized(s)

def normalized(s):
  # TODO: Test again on example sentences. Check if the ⟦re⟧ module works instead of ⟦regex⟧.
  assert isinstance(s, str)
  if s == "":
    return s
  # Normalizing nonstandard characters:
  s = s.strip()
  s = re.sub("[x’]", "'", s)
  s = re.sub("i", "ı", s)
  s = re.sub("ȷ", "j", s)
  s = re.sub("(?<=^)'", "", s)
  s = unicodedata.normalize("NFC", s)
  if None == re.search(
    "[\sáéíóúäëïöüâêîôû]", s, re.IGNORECASE
  ):
    # The input is a lemma.
    s = s[0] + s[1:].lower()
    s = unicodedata.normalize("NFC", s)
  else:
    # The input is a normal Toaq text or fragment (not a lemma form).
    s = re.sub("[\t ]+", " ", s)
    # Currently incorrect capitalization is not corrected.
    p = ( "\\b" )
    l = re.split(p, s)
    def f(w):
      p = "^(?:[" + std_initial_str + "]h?)?([" + std_vowel_str + "])"
      r = re.findall(p, w, re.IGNORECASE)
      if r != []:
        main_vowel_pos = w.index(r[0])
        bare = _with_replaced_characters(
          w,
          "áéíóúäëïöüâêîôû",
          "aeıouaeıouaeıou")
        if bare in toneless_particles:
          return bare
        elif bare in functors_with_grammatical_tone:
          pass
        v = w[main_vowel_pos]
        w = bare[:main_vowel_pos] + v + bare[main_vowel_pos + 1:]
      return w
    i = 1
    while i < len(l):
      l[i] = f(l[i])
      i += 2
    s = "".join(l)
  return s

def inflected_from_lemma(lemma, tone):
  assert is_a_lemma(lemma)
  assert not lemma in functors_with_lexical_tone | interjections
  i = _first(lemma, lambda c: c in "aeıou")
  lemma = _with_replaced_interval(
    lemma, i, i + 1, inflected_vowel(lemma[i], tone))
  return lemma

def inflected_vowel(vowel, tone):
  assert vowel in "aeıou"
  if tone in {"´", "́"}:
    targets = "áéíóú"
  elif tone in {"^", "̂"}:
    targets = "âêîôû"
  elif tone in {"¨", "̈"}:
    targets = "äëïöü"
  else:
    raise Exception(f"Invalid tone: ⟪{tone}⟫")
  return _with_replaced_characters(vowel, "aeıou", targets)

def _first(iterable, has_property):
    i = next(
      (i for i, e in enumerate(iterable) if has_property(e)),
      None)
    if i is None:
      raise ValueError
    else:
      return i

def _with_replaced_interval(s1, i, j, s2):
  assert isinstance(s1, str)
  assert isinstance(s2, str)
  assert isinstance(i, int)
  assert isinstance(j, int)
  assert i < j
  return s2.join([s1[:i], s1[j:]])


def is_an_inflected_contentive(s):
  return None != re.match(
    ( f"([{std_consonant_str}]h?)?"
    + f"[{std_vowel_str}]"
    + f"[aeıou]*[mq]?(([{std_consonant_str}]h?)[aeıouạẹı̣ọụ]+[mq]?)*$" ),
    s)

def is_a_contentive_lemma(s):
  return None != re.match(
    ( f"([{std_word_initial_str}]h?)?[aeıouạẹı̣ọụ]+[mq]?"
    + f"(([{std_consonant_str}]h?)[aeıouạẹı̣ọụ]+[mq]?)*$" ),
    s)

def is_a_lemma(s):
  return (
    is_a_contentive_lemma(s) or s in (
      toneless_particles | functors_with_lexical_tone | interjections))

def __is_an_interjection(s):
  return None != re.match(
      u"[áéíóúäëïöüâêîôû][aeiıou]*$", s)

# ==================================================================== #

File: nimtoaq/latin.nim
# LATIN TOAQ MODULE

# ==================================================================== #

import std/sets
from std/sequtils import mapIt
from std/tables import contains
import std/nre
from std/strutils import contains, find, join
from std/strformat import fmt
from std/unicode import
  `==`, `$`, Rune, toRunes, toUTF8, runeLen, runeAtPos
from std/enumerate import enumerate
from std/sugar import `=>`

proc `@&`[T](keys: openArray[T]): HashSet[T] =
  return sets.toHashSet(keys)


# ==================================================================== #

const vowel_str* = "aeiıouáéíóúäëïöüâêîôûạẹịı̣ọụạ́ẹ́ị́ọ́ụ́ạ̈ẹ̈ị̈ọ̈ụ̈ậệị̂ộụ̂"
const std_vowel_str* = "aeıouáéíóúäëïöüâêîôûạẹı̣ọụạ́ẹ́ị́ọ́ụ́ạ̈ẹ̈ị̈ọ̈ụ̈ậệị̂ộụ̂"
const vowels* = vowel_str
const std_vowels* = std_vowel_str
const consonant_str* = "'bcdfghjȷklmnprstzqꝡ"
const initial_str* = "'bcdfghjȷklmnprstzꝡ"
const word_initial_str* = "bcdfghjȷklmnprstzꝡ"
const std_consonant_str* = "'bcdfghjklmnprstzqꝡ"
const std_initial_str* = "'bcdfghjklmnprstzꝡ"
const std_word_initial_str* = "bcdfghjklmnprstzꝡ"
const charset* = vowel_str & consonant_str
const std_charset* = std_vowel_str & std_consonant_str

const initials* = @["m", "b", "p", "f", "n", "d", "t", "z", "c", "s", "r", "l", "nh", "j", "ȷ", "ch", "sh", "ꝡ", "g", "k", "'", "h"]
const finals* = @["m", "q"]
const consonants* = initials & @["q"]
const std_initials* = @["m", "b", "p", "f", "n", "d", "t", "z", "c", "s", "r", "l", "nh", "j", "ch", "sh", "ꝡ", "g", "k", "'", "h"]
const std_consonants* = std_initials & @["q"]

# ==================================================================== #

const matrix_subordinators* = @&["ꝡa", "ma", "tıo"]
const nominal_subordinators* =
   @&["ꝡä", "mä", "tïo", "lä", "ꝡé", "ná", "é"]
const adnominal_subordinators* = @&["ꝡë", "ë", "jü"]
const predicatizers* = @&["jeı", "mea", "po"]
const determiners* = @&[
  "ló", "ké", "sá", "sía", "tú", "túq", "báq", "já", "hí", "ní", "hú"]
const type_1_conjunctions* = @&["róı", "rú", "rá", "ró", "rí", "kéo"]
const type_2_conjunctions* = @&["roı", "ru", "ra", "ro", "rı", "keo"]
const type_3_conjunctions* = @&["rôı", "rû", "râ", "rô", "rî", "kêo"]
const conjunctions* = (
  type_1_conjunctions + type_2_conjunctions + type_3_conjunctions)
const falling_tone_illocutions* =
  @&["ka", "da", "ba", "nha", "doa", "ꝡo"]
const peaking_tone_illocutions* = @&["dâ", "môq"]
const raising_tone_illocutions* = @&["móq"]
const illocutions* = (
  falling_tone_illocutions + raising_tone_illocutions +
  peaking_tone_illocutions)
const focus_markers* = @&["kú", "tóu", "béı"]
const preverbals* = @&["bï", "nä"]
const vocative* = @&["hóı"]
const terminators* = @&["teo", "kı"]
const exophoric_pronouns* = @&[
  "jí", "súq", "nháo", "súna", "nhána", "úmo", "íme", "súho", "áma", "há"
]
const endophoric_pronouns* =
  @&["hóa", "hó", "máq", "tá", "hóq", "róu", "zé", "bóu", "áq", "chéq"]
const pronouns* = exophoric_pronouns + endophoric_pronouns

const functors_with_grammatical_tone* =
  predicatizers + @&["mı", "shu", "mo"]

const functors_with_lexical_tone* = (
  matrix_subordinators + nominal_subordinators +
  adnominal_subordinators + determiners + conjunctions + illocutions +
  focus_markers + preverbals + vocative + terminators + @&["gö", "kïo"]
)

const functor_lemmas* = (
  functors_with_grammatical_tone + functors_with_lexical_tone)

const toneless_particles* = (
  matrix_subordinators + falling_tone_illocutions + terminators)

const interjections* = @&[
  "m̄", "ḿ", "m̈", "m̉", "m̂", "m̀", "m̃"
]

# ==================================================================== #

func is_a_contentive_lemma*(s: string): bool =
  return isSome(nre.match(s, re(
    "(*UTF8)" &
    fmt"([{std_word_initial_str}]h?)?[aeıouạẹı̣ọụ]+[mq]?" &
    fmt"(([{std_consonant_str}]h?)[aeıouạẹı̣ọụ]+[mq]?)*$"
  )))

func is_a_lemma*(s: string): bool =
  return (
    is_a_contentive_lemma(s) or
    (toneless_particles + functors_with_lexical_tone + interjections)
    .contains(s)
  )

func with_replaced_interval[T](s1: T, i: int, j: int, s2: T): T
func first[T](
  s: seq[T], has_property: proc (e: T): bool {.noSideEffect.}): int
func inflected_vowel(vowel: string, tone: string): string
func utf8_with_replaced_characters[T: seq[Rune] or string](
  str: T,
  src_chars: string,
  rep_chars: string
): string

func inflected_from_lemma*(lemma: string, tone: string): string =
  assert(runeLen(tone) == 1)
  assert(is_a_lemma(lemma))
  assert(
    not (functors_with_lexical_tone + interjections).contains(lemma))
  let lemma_rs = toRunes(lemma)
  const vowels = toRunes("aeıou")
  let i = first(lemma_rs, (rune: Rune) => vowels.contains(rune))
  return (
    with_replaced_interval(
    lemma_rs, i, i + 1, toRunes(
      inflected_vowel(lemma_rs[i].toUTF8(), tone)))
  ).mapIt(it.toUTF8()).join()

func inflected_vowel(vowel: string, tone: string): string =
  assert("aeıou".contains(vowel))
  var targets: string
  if @["´", "́"].contains(tone):
    targets = "áéíóú"
  elif @["^", "̂"].contains(tone):
    targets = "âêîôû"
  elif @["¨", "̈"].contains(tone):
    targets = "äëïöü"
  else:
    raise newException(ValueError, fmt"Invalid tone: ⟪{tone}⟫")
  return utf8_with_replaced_characters(vowel, "aeıou", targets)

func first[T](
  s: seq[T],
  has_property: proc (e: T): bool {.noSideEffect.}
): int =
  for i, e in enumerate(s):
    if has_property(e):
      return i
  return -1

func first_rune(
  s: string,
  has_property: proc (rune: Rune): bool {.noSideEffect.}
): int =
  let l = runeLen(s)
  for i in 0 ..< l:
    if has_property(runeAtPos(s, i)):
      return i
  return -1

func with_replaced_interval[T](s1: T, i: int, j: int, s2: T): T =
  assert(i < j)
  return s1[ 0..< i] & s2 & s1[j .. ^1]

func utf8_with_replaced_characters[T: seq[Rune] or string](
  str: T,
  src_chars: string,
  rep_chars: string
): string =
  let src_l = toRunes(src_chars)
  let rep_l = toRunes(rep_chars)
  when T is string:
    var str_l = str
  else: # T is seq[Rune]
    var str_l = toRunes(str)
  var i = 0
  while i < str_l.len:
    if src_l.contains(str_l[i]):
      var k = first(src_l, (r: Rune) => r == str_l[i])
      str_l = str_l[0 ..< i] & rep_l[k] & str_l[i + 1 .. ^1]
    i += 1
  return str_l.mapIt($it).join("")

